rfcs/20230420-kernel-cache/README.md

```diff
@@ -0,0 +1,292 @@
+# Reducing Primitive Creation Overhead for Dynamic Shapes
+
+## Motivation
+
+oneDNN API requires users to specify tensors dimensions to create a primitive,
+which makes the primitive dependent on the tensors's shapes. Primitive creation is
+an expensive operation, this is why oneDNN has a primitive cache to reduce the
+overhead. However, the primitive cache can only help if the tensors shapes do not
+change between workload iterations otherwise the primitives have to be created for
+the new shapes which incurs additional overhead. The goal of this RFC is to propose
+a solution to reduce that overhead.
+
+## Background
+
+There is a reason why the oneDNN API requires users to specify tensors dimensions
+to create a primitive. The library may use the information to generate a kernel
+specifically for the given shapes to get the best performance.
+
+However, the information may not always be used by the library meaning that the
+generated kernels can be used for different shapes, which makes requirement to
+always specify tensors dimensions is overly strict.
+
+There was an attempt in the past to relax the requirement by introducing so-called
+runtime dimensions support. The idea was to allow users specify a placeholder
+(`DNNL_RUNTIME_DIM_VAL`) instead of the actual dimensions. The general rule for
+the feature was that the more actual dimensions users provide the better
+performance they can get.
+
+It's obvious that in order to support runtime dimensions the kernels must not
+depend on the shapes. Or there should be different kernels that are specific to
+the amount of information provided. And the kernels should also demonstrate
+reasonably good performance otherwise, the creation overhead will be replaced by
+a suboptimal performance overhead, which will defeat the purpose of having the
+runtime dimensions feature.
+
+The runtime dimensions feature is not a silver bullet, this is why the RFC will
+consider another option to reduce the primitive creation overhead.
+
+## Proposal
+
+There are two options to reduce primitive creation overhead:
+1. Extend the runtime dimensions feature for the relevant primitives
+2. Introduce a cache for kernels
+
+### Option 1: Cache for Kernels (recommended)
+
+The idea is to introduce a cache for kernels (hereinafter *cache*). The cache will
+be global similar to the primitive cache.
+
+Pros:
+* Doesn't require users to modify their code to make the feature work
+* Eliminates the kernel creation overhead completely when it's possible
```

Comment:
Whether the kernel creation overhead can be elimination for such case depends on whether the kernel can be re-uses for those different shapes. If it can't then there is nothing we can do about it. This is true for all options.

Reply:
Then, for such kind case, how many shape variants can be covered by one kernel should be the final solution, instead of a kernel cache?

Reply:
The problem is that it's always possible to cover different shapes with a single kernel. For example, a hypothetical primitive may require tens of kernels to support all shapes. If we created all of them during the primitive creation it would take unreasonable amount of time. If we don't then we can't fully support runtime dimensions for the primitive. There is an option to reduce amount of kernels by supporting only a subset of dimensions combinations that can be runtime but it will come with another set of problem that I mentioned in the "cons" section of the runtime dimensions option.

Reply:
I might not foresee whether the RFC could help us to resolve dynamic shape in our AI workload.
The RFC explains an adjustment of primitive creation without changing kernel generation heuristic, to share kernels with possible shapes. I have no idea whether there is overhead reduction in our dynamic shape case (whether these shapes can share same kernel). We may provide some real shapes to you.
I might think, to generate general kernel with offloading dynamic dimension to control loops dynamically that you mentioned as another option might be more beneficial for our cases. General kernel should be more compatible compared against existing JIT kernel for static dimension. BTW, heuristic to trad-off performance between general kernel with dynamic dimension and JIT kernel with static dimension need to be considered.